runtime/ops/io.rs

```diff
@@ -460,153 +530,163 @@ pub fn op_read(
 
   if is_sync {
-    MinimalOp::Sync({
-      // First we look up the rid in the resource table.
-      std_file_resource(&mut state.borrow_mut(), rid as u32, move |r| match r {
-        Ok(std_file) => {
-          use std::io::Write;
-          std_file
-            .write(&zero_copy[0])
-            .map(|nwritten: usize| nwritten as i32)
-            .map_err(AnyError::from)
-        }
-        Err(_) => Err(type_error("sync read not allowed on this resource")),
-      })
-    })
+    MinimalOp::Sync(op_write_sync(state, rid, buf))
   } else {
-    let zero_copy = zero_copy[0].clone();
-    MinimalOp::Async({
-      async move {
-        let resource = state
-          .borrow()
-          .resource_table
-          .get_any(rid as u32)
-          .ok_or_else(bad_resource_id)?;
-        let nwritten = if let Some(stream) =
-          resource.downcast_rc::<TcpStreamResource>()
-        {
-          stream.write(&zero_copy).await?
-        } else if let Some(stream) = resource.downcast_rc::<StreamResource>() {
-          stream.clone().write(&zero_copy).await?
-        } else {
-          return Err(bad_resource_id());
-        };
-        Ok(nwritten as i32)
-      }
-      .boxed_local()
-    })
+    MinimalOp::Async(op_write_async(state, rid, buf).boxed_local())
   }
 }
 
-pub fn std_file_resource<F, T>(
-  state: &mut OpState,
-  rid: u32,
-  mut f: F,
-) -> Result<T, AnyError>
-where
-  F: FnMut(Result<&mut std::fs::File, ()>) -> Result<T, AnyError>,
-{
-  // First we look up the rid in the resource table.
+fn op_write_sync(
+  state: Rc<RefCell<OpState>>,
+  rid: i32,
+  buf: ZeroCopyBuf,
+) -> Result<i32, AnyError> {
+  let rid = rid.try_into().map_err(|_| bad_resource_id())?;
+  StdFileResource::with(&mut state.borrow_mut(), rid, move |r| match r {
+    Ok(std_file) => std_file
+      .write(&buf)
+      .map(|nwritten: usize| nwritten as i32)
+      .map_err(AnyError::from),
+    Err(_) => Err(not_supported()),
+  })
+}
+
+async fn op_write_async(
+  state: Rc<RefCell<OpState>>,
+  rid: i32,
+  buf: ZeroCopyBuf,
+) -> Result<i32, AnyError> {
+  let rid = rid.try_into().map_err(|_| bad_resource_id())?;
   let resource = state
+    .borrow()
     .resource_table
-    .get::<StreamResource>(rid)
+    .get_any(rid)
     .ok_or_else(bad_resource_id)?;
+  let nwritten = if let Some(s) = resource.downcast_rc::<ChildStdinResource>() {
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<TcpStreamResource>() {
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<TlsClientStreamResource>() {
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<TlsServerStreamResource>() {
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<UnixStreamResource>() {
+    s.write(&buf).await?
+  } else if let Some(s) = resource.downcast_rc::<StdFileResource>() {
+    s.write(&buf).await?
+  } else {
+    return Err(not_supported());
+  };
```

Comment:
I like that the conditional is now explicitly expressed here in the op. Much easier to read.